### PR TITLE
liberror: introduce a TSS2_RC error decoding library

### PR DESCRIPTION
**Pull request by williamcroberts**
Add "liberror" which is a TSS2_RC error decoding library
analogous to C's strerror(3).

The TSS2_RC code contains a layer indicator field. This field
is used to inidcate what layer issued the error. Allow the
setting of custom error layer handlers so users can add
decoders for layer specific errors.

Add unit tests and package config information.

Signed-off-by: William Roberts <william.c.roberts@intel.com>

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -28,13 +28,15 @@
...
+INCLUDE_DIRS    = -I$(srcdir)/include -I$(srcdir)/sysapi/include -I$(srcdir)/marshal \
+    -I$(srcdir)/error
+
 ACLOCAL_AMFLAGS = -I m4
 AM_CFLAGS       = $(INCLUDE_DIRS) $(EXTRA_CFLAGS) $(CODE_COVERAGE_CFLAGS)
 AM_LDFLAGS      = $(EXTRA_LDFLAGS) $(CODE_COVERAGE_LIBS)
 
 # stuff to build, what that stuff is, and where/if to install said stuff
-lib_LTLIBRARIES = $(libmarshal) $(libsapi) $(libtcti_device) $(libtcti_socket)
+lib_LTLIBRARIES = $(liberror) $(libmarshal) $(libsapi) $(libtcti_device) $(libtcti_socket)
```
Comment (williamcroberts) [inline Makefile.am]: overlength? fold?

File: error/error_handler.c [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdbool.h>
```
Comment (williamcroberts) [inline error/error_handler.c]: sort

File: error/error_handler.c [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+/**
+ * Clears out a static buffer by setting index 0 to the null byte.
+ * @param buffer
+ *  The buffer to clear out.
+ */
+void
+clearbuf (
+    char *buffer)
+{
+
```
Comment (williamcroberts) [inline error/error_handler.c]: extra newline here and throughout

File: error/error_handler.c [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    ADD_NULL_HANDLER,                       // layer 6  is the feature rc
+    ADD_HANDLER("fapi", NULL),              // layer 7  is the esapi rc
+    ADD_HANDLER("sys", sys_err_handler),    // layer 8  is the sys rc
+    ADD_HANDLER("mu",  sys_err_handler),    // layer 9  is the mu rc
+				            // Defaults to the system handler
+    ADD_HANDLER("tcti", sys_err_handler),   // layer 10 is the tcti rc
+					    // Defaults to the system handler
+    ADD_HANDLER("rmt", tpm2_error_handler), // layer 11 is the resource manager TPM RC
+				            // The RM usually duplicates TPM responses
+				            // So just default the handler to tpm2.
```
Comment (williamcroberts) [inline error/error_handler.c]: align

File: error/error_handler.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#ifndef ERROR_ERROR_HANDLER_H_
+#define ERROR_ERROR_HANDLER_H_
```
Comment (williamcroberts) [inline error/error_handler.h]: extra ERROR

File: error/tpm2_error.c [hunk 6]
```diff
@@ -0,0 +1,291 @@
+#include "../error/tpm2_error.h"
```
Comment (williamcroberts) [inline error/tpm2_error.c]: should be in local defines block, no relative path (dang you eclipse)
Also needs copyright block

File: error/tpm2_error.h [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#ifndef ERROR_TPM2_ERROR_H_
```
Comment (williamcroberts) [inline error/tpm2_error.h]: copyright block

File: lib/error.pc.in [hunk 8]
```diff
@@ -0,0 +1,6 @@
+Name: sapi
+Description: TSS2 error decoding library library.
```
Comment (williamcroberts) [inline lib/error.pc.in]: remove double library

File: error/error_handler.h [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Concatenates (safely) onto a static buffer given a format and varaidic
+ * arguments similar to sprintf.
+ * @param b
+ *   The static buffer to concatenate onto.
+ * @param fmt
+ *   The format specifier as understood by printf followed by the variadic
+ *   parameters for the specifier.
+ */
+#define catbuf(b, fmt, ...) _catbuf(b, sizeof(b), fmt, ##__VA_ARGS__)
```
Comment (tstruk) [inline error/error_handler.h]: Why do we need this wrapper? Can the _catbuf be renamed to catbuf?

File: error/error_handler.c [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    if (lname)
+        catbuf(buf, "%s:", lname);
+    else
+        catbuf(buf, "%u:", layer);
+
+    handler = !handler ? unkown_layer_handler : handler;
+
+    // Handlers only need the error bits. This way they don't
+    // need to concern themselves with masking off the layer
+    // bits or anything else.
```
Comment (tstruk) [inline error/error_handler.c]: The comment style is inconsistent. We are moving towards C style comments so C++ style shouldn't be used.

File: error/error_handler.h (hunk 9 above)
Comment (williamcroberts) [reply error/error_handler.h]: This wrapper exists so you don't have to do
```
 catbuf(buf, sizeof(buf), "%s", "foo")
```
You can drop the sizeof() parameter:
```
 catbuf(buf, "%s", "foo")
```
Easier to read IMO, but not married to it, we can just do sizeof().

File: error/error_handler.c (hunk 10 above)
Comment (williamcroberts) [reply error/error_handler.c]: Will fix.

**Comment by williamcroberts** [comment]
Moving back into tools.

**Pull request closed without merging**